src/main/java/github/alecsio/mmceaddons/common/assembly/AdvancedMachineAssembly.java

```diff
@@ -0,0 +1,279 @@
+import moze_intel.projecte.api.capabilities.IKnowledgeProvider;
+import moze_intel.projecte.utils.EMCHelper;
+import net.minecraft.block.state.IBlockState;
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.init.SoundEvents;
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.EnumFacing;
+import net.minecraft.util.EnumHand;
+import net.minecraft.util.SoundCategory;
+import net.minecraft.util.Tuple;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.text.TextComponentTranslation;
+import net.minecraft.world.World;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.common.util.BlockSnapshot;
+import net.minecraftforge.event.ForgeEventFactory;
+import net.minecraftforge.event.world.BlockEvent;
+import net.minecraftforge.fml.common.Optional;
+import net.minecraftforge.fml.common.gameevent.TickEvent;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import stanhebben.zenscript.annotations.NotNull;
+
+import java.util.*;
+
+/**
+ * Represents and encapsulates all the logic related to the assembly of a multiblock.
+ * Assemblies are managed by {@link github.alecsio.mmceaddons.common.assembly.MachineAssemblyManager}, which acts as
+ * a cache for and provider of scheduled assemblies.
+ * Assemblies are added to the manager from {@link ItemAdvancedMachineAssembler#onItemUse(EntityPlayer, World, BlockPos, EnumHand, EnumFacing, float, float, float)}
+ * and an assembly step is triggered every player tick from {@link MachineAssemblyEventHandler#onPlayerTick(TickEvent.PlayerTickEvent)}
+ */
+public class AdvancedMachineAssembly extends AbstractMachineAssembly {
+
+    private static final Logger LOGGER = LogManager.getLogger();
+
+    // The last error that was reported by any of the handlers. If multiple errors were reported, only the last
+    // one will be displayed to the player.
+    private TextComponentTranslation lastError;
+
+    public AdvancedMachineAssembly(World world, BlockPos controllerPos, EntityPlayer player, StructureIngredient ingredient) {
+        super(world, controllerPos, player, ingredient);
+    }
+
+    /**
+     * Processes up to {@link MMCEAConfig#blocksProcessedPerTick} blocks from the assembly
+     */
+    @Override
+    public void assembleTick() {
+        List<StructureIngredient.ItemIngredient> itemIngredients = ingredient.itemIngredient();
+        Iterator<StructureIngredient.ItemIngredient> iterator = itemIngredients.iterator();
+
+        for (int i = 0; i < MMCEAConfig.blocksProcessedPerTick; i++) {
+            if (!iterator.hasNext()) {
+                completed = true;
+                break;
+            }
+            StructureIngredient.ItemIngredient ingredientToProcess = iterator.next();
+            tryPlaceBlock(ingredientToProcess);
+            iterator.remove();
```

Comment:
Does this have null safety? is it possible that tryPlaceBlock fails, if then, does it stop building or expected to continue?